Control.cpp

```diff
@@ -0,0 +1,492 @@
+  double error = current_distance - desire;
+  return error;
+}
+
+void Control::getPwm(double &vel) { 
+  if (vel < kLimit_inf_pwm) {
+  vel = kLimit_inf_pwm;
+  }
+  
+  if (vel > kLimit_sup_pwm) {
+    vel = kLimit_sup_pwm;
+  }
+}
+
+double Control::getDistanceFrontLeft() {
+  delay(50);
+  return sonar[0].ping_cm();
+}
+
+double Control::getDistanceFrontRight() {
+  delay(50);
+  return sonar[1].ping_cm();
+}
+
+double Control::getDistanceRightUp() {
+  delay(50);
+  return sonar[2].ping_cm();
+}
+
+double Control::getDistanceRightDown() {
+  delay(50);
+  return sonar[3].ping_cm();
+}
+
+double Control::getDistanceLeftUp() {
+  delay(50);
+  return sonar[4].ping_cm();
+}
+
+double Control::getDistanceLeftDown() {
+  delay(50);
+  return sonar[5].ping_cm(); 
+}
+
+bool Control::checkWallsRight() {
+  double current_distance_right_up = getDistanceRightUp();
+  double current_distance_right_down = getDistanceRightDown();
+  if (current_distance_right_up < 15 and current_distance_right_down < 15) {
+    return true;
+  }
+  else if (current_distance_right_up > 15 || current_distance_right_down > 15) {
+    return false;
+  }
+}
+
+bool Control::checkWallsLeft() {
+  double current_distance_left_up = getDistanceLeftUp();
+  double current_distance_left_down = getDistanceLeftDown();
+  if (current_distance_left_up < 15 and current_distance_left_down < 15) {
+    return true;
+  }
+  else if (current_distance_left_up > 15 || current_distance_left_down > 15) {
+    return false;
+  }
+}
+
+void Control::advance(const double desire, const double desireUltrasonic) {
+  double encoder1 = 0;
+  double encoder2 = 0;
+  double pwm_left = 0;
+  double pwm_right = 0;
+  double final_error = 0;
+  double current_distance_right_up = getDistanceRightUp();
+  double current_distance_right_down = getDistanceRightDown();
+  double current_distance_left_up = getDistanceLeftUp();
+  double current_distance_left_down = getDistanceLeftDown();
+  double error_right_up = getErrorUltrasonic(current_distance_right_up, desireUltrasonic);
+  double error_right_down = getErrorUltrasonic(current_distance_right_down, desireUltrasonic);
+  double error_left_up = getErrorUltrasonic(current_distance_left_up, desireUltrasonic);
+  double error_left_down = getErrorUltrasonic(current_distance_left_down, desireUltrasonic);
+  double current_angle = getAngleBNOX();
+  double errorBNO = getError(current_angle, desire);
+
+  do {
+    double error_ultrasonic_right = error_right_up + error_right_down;
+    double error_ultrasonic_left = error_left_up + error_left_down;
+    if (error_ultrasonic_left > 0) {
+      error_ultrasonic_left = -(error_ultrasonic_left);
+    }
+    else {
+      error_ultrasonic_left = -(error_ultrasonic_left);
+    }
+    final_error = errorBNO + error_ultrasonic_right + error_ultrasonic_left;
+    if (final_error > 0) {
+    pwm_right = kLimit_inf_pwm;
+    pwm_left = kP2 * final_error;
+    getPwm(pwm_left);
+  }
```

Comment:
Maybe you want to calculate pwm_right_bno and pwm_right_sonar differently. In other words, do not combine them in a final_error.